packages/runtime/src/helpers/redirects.ts

```diff
@@ -191,13 +248,14 @@ const generateDynamicRewrites = ({
   const dynamicRewrites: NetlifyConfig['redirects'] = []
   const dynamicRoutesThatMatchMiddleware: Array<string> = []
   dynamicRoutes.forEach((route) => {
-    if (isApiRoute(route.page)) {
+    if (isApiRoute(route.page) || is404Route(route.page, i18n)) {
       return
     }
     if (route.page in prerenderedDynamicRoutes) {
+      const { fallback } = prerenderedDynamicRoutes[route.page]
       if (matchesMiddleware(middleware, route.page)) {
         dynamicRoutesThatMatchMiddleware.push(route.page)
-      } else {
+      } else if (fallback !== false) {
```

Comment:
this is the important bit as it allows non-prerendered dynamic routes with `fallback: false` to fall through and hit the 404 redirects